nexus/reconfigurator/planning/src/planner.rs

```diff
@@ -972,58 +1014,118 @@ impl<'a> Planner<'a> {
-            let num_zones_to_add =
-                self.num_additional_zones_needed(zone_kind, report);
-            if num_zones_to_add == 0 {
-                continue;
-            }
-            // We need to add at least one zone; construct our `zone_placement`
-            // (or reuse the existing one if a previous loop iteration already
-            // created it).
-            let zone_placement = zone_placement.get_or_insert_with(|| {
-                // This constructs a picture of the sleds as we currently
-                // understand them, as far as which sleds have discretionary
-                // zones. This will remain valid as we loop through the
-                // `zone_kind`s in this function, as any zone additions will
-                // update the `zone_placement` heap in-place.
-                let current_discretionary_zones = self
-                    .input
-                    .all_sled_resources(SledFilter::Discretionary)
-                    .filter(|(sled_id, _)| {
-                        !report.sleds_waiting_for_ntp_zone.contains(&sled_id)
-                    })
-                    .map(|(sled_id, sled_resources)| {
-                        OmicronZonePlacementSledState {
-                            sled_id,
-                            num_zpools: sled_resources
-                                .all_zpools(ZpoolFilter::InService)
-                                .count(),
-                            discretionary_zones: self
-                                .blueprint
-                                .current_sled_zones(
-                                    sled_id,
-                                    BlueprintZoneDisposition::is_in_service,
-                                )
-                                .filter_map(|zone| {
-                                    DiscretionaryOmicronZone::from_zone_type(
-                                        &zone.zone_type,
+            // Our goal here is to make sure that if we have less redundancy for
+            // discretionary zones than needed, we deploy additional zones.
+            //
+            // For most zone types, we only care about the total count of that
+            // kind of zone, regardless of image. In contrast, for Nexus, we may
+            // need to reach a minimum redundancy count for multiple zone images
+            // (new and old) during a handoff.
+            let image_sources = match zone_kind {
+                DiscretionaryOmicronZone::Nexus => {
+                    let new_repo = self.input.tuf_repo().description();
+                    let new_image =
+                        new_repo.zone_image_source(zone_kind.into())?;
+
+                    let mut images = vec![];
+                    let Some(nexus_in_charge_image) =
+                        self.lookup_current_nexus_image()
+                    else {
+                        return Err(Error::NoActiveNexusZonesInParentBlueprint);
+                    };
```

Comment:
nit: I'd have `lookup_current_nexus_image()` return this error.  (This caller is inferring that's the only reason this could return `None`.)